PSR-3 compliance

enov opened the pull request:
This is a work-in-progress for PSR-3 compliance.

enov committed:
Declare required logger interfaces

Adds a Kohana_Logger interface that basically extends the PRS-3
LoggerInterface, with additional methods for deffered logging.

At this point any implementation of Kohana_Logger interface is
tied to the Log_Writer class. We can later decouple it by
introducing a separate interface to abstract the Log_Writer.

Also adds a Logger interface by basically extends Kohana_Logger.
This is just for the CFS enhanced extension, to keep the trend
with other interfaces in Kohana. IMHO, however, I can not
understand the value for this, as basically, a tweeked interface
can make more harm than any benefit.

enov committed:
Adds a PSR-3 compliant text interpolation function to Text class

Text interpolation in Kohana is done through `strtr` by directly
interpolating provided variables.

PSR-3 however, dictates on us to interpolate by enclosing the
variables with curly brackets in the message string.

IMHO, for consistency, we should also interpolate variables in
messages and I18n the same way we are interpolating log messages

enov committed:
PSR-3 compliance

Move logging logic into `log` method.

Deprecate `add` method, use `log` method within `add`
by sanitizing parameters for PSR-3 compliance

Update `Log_Writer` and `Log_Syslog` to match new string
levels coming from `Log` messages

enov commented:
Hmm, this is not based on 3.4/develop. It seems to me that I have created a branch starting from the wrong revision.

enov commented:
Or maybe I did not pull before starting.

enov committed:
Require psr/log in composer.json

enov committed:
Use Log::log or equivalent instead of Log::add

Now that `Log::add` is deprecated, use `Psr\Log\LoggerInterface`
methods instead, Log::log or the equivalent, appropriate methods

Also, when we are forcing the logs to be written, we are
testing if `Kohana::$log` is instanceof Logger. Otherwise
the `write` method might not be available in other logging
systems that are still compatible with `Psr\Log\LoggerInterface`

enov committed:
Relax log requirement constraint

We do not need to force `Kohana::$log` to be instanceof `Log`
anymore as the core is now Psr\Log\LoggerInterface compatible

enov commented:
@acoulton, could you kindly review this? Also, kindly guide me what types of tests I need to write to make this complete?

cyrgit committed:
Adds PHPDoc to Logger::log

cyrrill commented:
This is freking great, I'm very glad to see Kohana move closer to full PSR :) :+1:

enov committed:
Use Psr\Log\InvalidArgumentException for an invalid level

Check if level is available, acceptable levels are the strings
defined in `Psr\Log\LogLevel` as well as the int constants
defined in `Log`.

Now that int is an acceptable, I have removed the int-to-string
level conversion from `Log::add`

enov committed:
Cast $message argument to string in compliance with PSR-3

enov committed:
Adds usage examples to Log::log and minor fixes to docs

acoulton commented:
@enov this is looking great. A couple of thoughts:
- Do we need a Logger and Kohana_Logger interface? What benefit is there to someone being able to transparently extend the interface?
- Might it be worth introducing a class (perhaps with ArrayAccess for compatibility) for the log message itself, rather than the simple array that Log::log pushes onto the _messages array? Would allow better typehinting for that property and also for writer methods etc.

In terms of testing, things I'd consider covering (I've not reviewed what if anything is covered by existing tests):

#### Kohana_Core::init
- Creates default log instance if not set
- Does not change log instance if already assigned to a dummy \Psr\Log\LoggerInterface

#### Kohana_Exception::log
- Behaviour when Kohana::$log is not assigned (fails silently?)
- Logs the exception
- Writes logs if a Logger
- Does not fail (eg from attempting to call write) if the Kohana::$log is a non-kohana LoggerInterface

### Kohana_Log
- I'd cover the inherited AbstractLogger methods for completeness, eg with a self-shunt extension class that captures calls to log and verifies that the level-specific methods call it as expected.
- That add calls log with expected argument structures
- How ::log handles the various types of levels (string, int, undefined)
- That ::log casts message to string
- That ::log interpolates context to message
- The various message properties are set in the messages array as expected
- That ::log writes if write_on_add is set - again could do this with a self-shunt

### Kohana_Log_Syslog
- That it calls syslog with expected arguments, eg using a self-shunt
- Level conversion
- Exception logging

And if Kohana_Log_Writer::format_message doesn't already have coverage then I guess also tests for that.

enov commented:
> Do we need a Logger and Kohana_Logger interface

No we don't, interfaces should not be transparently extended like implementations, IMHO.

> Might it be worth introducing a class for the log message itself

My initial intention was to be psr-3 compliant with the least change possible in the code base. I thought can further enhance the code in later versions. What do you say?

acoulton commented:
> > Do we need a Logger and Kohana_Logger interface
> 
> No we don't, interfaces should not be transparently extended like implementations, IMHO.

Agreed - but you have both in this PR. My preference would be to drop `Logger` and update the type checks throughout to `Kohana_Logger` since that keeps it clearer where that has come from (eg makes obvious it's not a PSR3 interface).

> > Might it be worth introducing a class for the log message itself
> 
> My initial intention was to be psr-3 compliant with the least change possible in the code base. I thought can further enhance the code in later versions. What do you say?

Fair point, yes this pull should be limited to what's strictly necessary for PSR3.

enov commented:
> My preference would be to drop Logger and update the type checks throughout to Kohana_Logger

Agreed. I will create an issue for this later, targeting all other interfaces that follow this pattern.

arteymix commented:
@enov do you think [PSR-7 http-message](https://github.com/php-fig/http-message) should be implemented as well?

enov commented:
@arteymix that's still under discussion. Plus, it would require us to be PSR-1/2 coding style, because unlike this PSR-3, it has long camelCased method names.

There is also a draft RFC about HTTP interface, that _might_ go into PHP 7:
https://wiki.php.net/rfc/http-interface

arteymix commented:
That's what I feared when I looked into the specification. PSR 1/2 is hardly thinkable at thins point.

It's a good thing though that we get more compliance with PSR, so I second you on that PR.

enov committed:
Remove Logger interface, use Kohana_Logger instead

interfaces should not be transparently extended like classes

enov commented:
@acoulton, I am struggling with writing the first test:

> Kohana_Core::init
> - Creates default log instance if not set

The problem is that PHPUnit boostraps Kohana the standard way, which calls `Kohana::init()`.

So, since  `Kohana::init()` is already called and log is set, in order for me to have an unset log, I have to unset it within the test. Now, while that seems to be easy, calling `Kohana::init()` again is not, because the protected flag `Kohana::$_init` that checks if it is initialized. That would lead us to call `Kohana::deinit()` within the test, which sets that flag to `FALSE`, but would remove also the registration of the standard Kohana autoloader, which makes other tests to fatal.

It seems to me I am hunting some kind of chicken and egg problem. I found the following 2 sub-par solutions:

## Solution 1: Replay parts of bootstrap within the test

To be able to `deinit()` and then re`init()`, I need to replay, in between these calls, parts of bootstrap required for tests to continue. I think this is sub-par solution because changing the bootstrap later on, would require us to revisit these tests and fix them.

Kindly find here-under the code for this solution:

``` php
    /**
     * Tests Kohana::init() for default log file creation
     */
    public function test_init_creates_default_log_instance_if_not_set()
    {
        // de-initialize Kohana, should set Kohana::$log to NULL
        Kohana::deinit();

        // re-register autoloader as de-initialization removes that
        spl_autoload_register(array('Kohana', 'auto_load'));

        // test if Kohana::$log is NULL at this point
        $this->assertNull(Kohana::$log);

        // re-initialize Kohana
        Kohana::init();

        // test if a default log instance is created
        $this->assertInstanceOf('Log', Kohana::$log);

        /**
         * At this point we need to replay some part from bootstrap
         * so that other tests may continue normally
         */

        // Attach the file write to logging. Multiple writers are supported.
        Kohana::$log->attach(new Log_File(APPPATH . 'logs'));

        //Attach a file reader to config. Multiple readers are supported.
        Kohana::$config->attach(new Config_File);
    }
```

## Solution 2: a TestableCore that extends Core with a public set_init_flag

This solution relies on the fact that we can run `init()` again if the flag `$_init` is set to `FALSE`, which help us avoid to `deinit()` Kohana. I do think this solution is also sub-par, because it re-registers the exception handler, error handler and shutdown handler with PHP runtime again.

Please see the code for this solution here:

``` php
    /**
     * Tests Kohana::init() for default log file creation
     */
    public function test_init_creates_default_log_instance_if_not_set()
    {
        // set $log to NULL, as it inherited the instance from the parent class
        Kohana_CoreTest_TestableCore::$log = NULL;

        // test if it is NULL (do we need this?)
        $this->assertNull(Kohana_CoreTest_TestableCore::$log);

        // set the flag Kohana::$_init to FALSE to be able to run init() again
        Kohana_CoreTest_TestableCore::set_init_flag(FALSE);

        // initialize Kohana again
        Kohana_CoreTest_TestableCore::init();

        // assert
        $this->assertInstanceOf('Log', Kohana_CoreTest_TestableCore::$log);
    }
```

Code for `Kohana_CoreTest_TestableCore`:

``` php
class Kohana_CoreTest_TestableCore extends Kohana_Core {

    /**
     * Sets the init flag to force initialization/deinitialization
     * @param bool $flag
     */
    public static function set_init_flag($flag)
    {
        static::$_init = $flag;
    }

}
```

Thanks.

acoulton commented:
@enov yes, that's all fairly horrid isn't it :)

Not sure what to suggest, as you say the main issue is that `init` does so much and replaying all of that in the unit test is bound to cause problems.

I suppose you could split all the separate actions in init into protected methods, then stub a class that makes all of them no-op - for example:

``` php
class Kohana_Core {
  public static function init() {
    if (Kohana::$_init)
    {
      // Do not allow execution twice
      return;
    }

    // Kohana is now initialized
    Kohana::$_init = TRUE;
    static::init_profile();
    static::init_errors();
    static::init_shutdown();
    static::init_windows();
    static::init_cache();
    // etc
    static::init_log();
    static::init_config();
  }
}

class LogInitKohana_CoreStub extends Kohana_Core {
  public static function init()
  {
    self::$_init = FALSE;
  }

  public static function init_profile() {}
  public static function init_errors() {}
  public static function init_cache() {}
  // etc for everything except init_log
}
```

That's not very nice either because of the need to be sure every `init_` method was extended with an empty stub over time.

Another option would be to break the init method up into a stack of initialisers (classes or callables of some kind) - then you could just test the LogInitialiser. That would be cleaner and give more scope for extension in future and by end-users, but is obviously a chunky refactoring and may have performance, BC and other considerations.

Maybe for now the only sensible thing in the current architecture is to assume that PHPUnit didn't itself assign anything to `Kohana::$log` and so if it has a value it must have been assigned by `init` - in other words just test the result. It might be better than trying to sort out the whole init method as part of this feature.

enov commented:
> assume that PHPUnit didn't itself assign anything to Kohana::$log and so if it has a value it must have been assigned by init - in other words just test the result.

If we can get away with that for the first test, we can not for the second test you suggested :smile:

> Kohana_Core::init
> -   Does not change log instance if already assigned to a dummy \Psr\Log\LoggerInterface

acoulton commented:
@enov indeed. I think maybe we have to accept that's almost impossible to test without introducing a lot of changes that aren't directly relevant to this PR. The only other thing I thought of was shelling out to a script in a separate PHP process but that would be pretty unpleasant too. Maybe for now these are things that just have to be verified by eye/knowing they work in end-user apps.

enov committed:
Tests if Kohana::init() creates default log instance

enov committed:
Test if Kohana_Exception logs exceptions

with a standard Kohana_Logger, as well as with a PSR-3 compliant
logger that is not a Kohana_Logger

enov committed:
use array_merge in Log_Writer mock for completeness

enov committed:
add tests for generic log() and 8 level methods

Tests are failing for the time being because
the trace information within the log adds
an additional entry in case of specialized
`debug()`, `info()`... method calls vs
the generic `log()` calls

enov commented:
I added tests comparing messages resulting from the generic `log()` method, as well as the 8 specialized _level_ methods.

**Tests are currently failing**. This is because `Kohana_Log` adds _trace_ information to the message. And since there is an additional method call when one of the _level_ methods are used, we are having different _trace_ entries in those messages when we compare them to the message generated by the generic `log()` method.

PSR-3 specifies that:

> A ninth method, `log`, accepts a log level as first argument. Calling this method with one of the log level constants MUST have the same result as calling the level-specific method.

I think we either need to normalize the _trace_ entries, by carefully re-engineering the result of the call to [debug_backtrace](https://github.com/kohana/core/blob/3.4/feature/psr-3-compliance/classes/Kohana/Log.php#L274), or to forget about the _trace_ entry, as it is ignored by the abstract `Log_Writer`.

For the record, the [_file_, _line_, _class_, and _function_](https://github.com/kohana/core/blob/3.4/feature/psr-3-compliance/classes/Kohana/Log.php#L287-290) entries are also giving different results. Sometimes they are `NULL`. I am not sure why. I also did some tests without [array_slice](https://github.com/kohana/core/blob/3.4/feature/psr-3-compliance/classes/Kohana/Log.php#L274), and results seemed to be slightly better (no `NULL`s, but again different results). I am even unsure why that `array_slice` is there, to begin with!

Any insight and guidance is greatly appreciated.

enov committed:
Remove *trace*, *class*, and *function* from log message

Removed *trace*, *class*, and *function* from log message, because we
can not have the "same log message results" when logs are generated from
*level* specific methods (`emergency`, `debug`, ...) vs the generic
`log` method.

It is worthy to note that the above elements are not used by the
`abstract class Kohana_Log_Writer`, so they are not usually logged,
unless Kohana_Log_Writer is extended by userland, using CFS.

Added a fuzzy `assertLogMessageEquals` to the test class to make log
message testing easier, because generating log messages always include
the line number from which the `log` (or equivalent *level* methods) are
called.

enov committed:
Add a comment to explain why a trace is conditionally removed

enov commented:
Alright, played with fire and removed _trace_, _class_, and _function_ keys from log message.

Please review.

cc @acoulton, @lenton, @rjd22, @shadowhand

enov commented:
Do **not** merge yet.

enov committed:
`assertLogMessageEquals`: add @codingStandardsIgnore + doc line

[ci skip]

enov commented:
@acoulton could you please review this? Specifically this commit: https://github.com/kohana/core/commit/32247e58cd3e43ea7e2eb6dbb56a5b8712639583

enov commented:
@acoulton I know you might be busy, but I just need your confirmation in case we can get rid of _trace_, _class_, and _function_ from log message (as we are not using them in the default log writer).

Those are still available in case of an exception. I'll write more tests to prove the case.

acoulton commented:
@enov so sorry, I've been flat out. I've just had a quick look over this and it's looking good.

I think it's fine to drop trace, class and function from ordinary log messages - as you say they're not written by any of our own writers and anyway that feels like something where the end-user should pass them as part of the context if they want them. Apart from unexpected exceptions (which are handled separately) usually you know the trace of application-specific logs.

enov commented:
Thanks for the review @acoulton .

rjd22 commented:
@enov I also reviewed it and it looks good. I'm just wondering if we should still do our own logging. Wouldn't it be better to move to monolog? What do you think @acoulton 

Anyway all in good time. For now :+1:

enov commented:
Thanks for the review @rjd22.

rjd22 commented:
@enov any updates on this?

enov commented:
@rjd22 Need to find time to write the remaining tests, and write some docs about how to plug-in a PSR-3 compatible logger, instead of using the stock Kohana logger.

rjd22 commented:
@enov Another poke :P. It would be really nice to have this in Kohana :)

enov commented:
I will try to concentrate my time on this :)
On 7 Oct 2015 13:21, "Robert-Jan de Dreu" notifications@github.com wrote:

> @enov https://github.com/enov Another poke :P. It would be really nice
> to have this in Kohana :)
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/kohana/core/pull/563#issuecomment-146144265.

enov committed:
Test that `Kohana::init()` creates default log instance if not set

And that it does not change log instance if already assigned to a
`\Psr\Log\LoggerInterface`.

Those tests are run via a separate PHPUnit process using the
@runInSeparateProcess annotation.

enov committed:
Test Kohana_Exception::log fails silently when Kohana::log is NULL

If Kohana_Exception::log fails with an exception, the test would not pass
https://github.com/sebastianbergmann/phpunit-documentation/issues/171

enov committed:
Test \Psr\Log\AbstractLogger

enov committed:
Rename test method and add a DocBlock comment to provider_logging

[ci skip]

enov committed:
Add unit test for Syslog

enov committed:
Support objects/arrays/resources a bit better

Better message interpolation from Jordi Boggiano
Shamelessly copied from the Monolog project:
https://github.com/Seldaek/monolog/commit/b3f039339d7a8173c7b441dbaa572ccacb712b54

enov committed:
Test that Log casts message to string

enov committed:
Extend the official tests \Psr\Log\Test\LoggerInterfaceTest

enov committed:
Refactor public static $write_on_add to protected $write_immediately

Rename variable because method `add` is now deprecated.
Change variable from public static to protected non-static.
Add getter/setter.

enov committed:
Test that Log writes immediately when immediate writing is ON

enov committed:
Fix test group name of official PSR logger test

enov committed:
Test log message arrays have the expected elements

enov committed:
Fix doc block of Low_Writer::$strace_level

enov committed:
Rename argument of `test_log_message` from `$method` to `$level`

enov committed:
Test formatting of log entries

enov committed:
Relax $time_delta when asserting fuzzily in assertLogMessageEquals

HHVM tends to be very slow when testing

enov commented:
I believe I have completed the writing of the requested tests, as well as implementing the official `\Psr\Log\Test\LoggerInterfaceTest`. Some tests seems to be overlapping. Please let me know if I am missing something.

Regarding the failure on PHPUnit v4.0.0, that's because tests in acd9d16ed5686207598c60823afe6d64c989bd4d are running via a separate PHPUnit process using both `@runInSeparateProcess` and `@preserveGlobalState disabled` annotations. When both are present, PHPUnit [prior to version 4.3.0](https://github.com/sebastianbergmann/phpunit/commit/7d8da4174961b3144e364990a499efe90c5b49f5) used to ignore the bootstrapping. That was fixed in https://github.com/sebastianbergmann/phpunit/pull/1364.

Kindly requesting your review and comments. Docs should be updated once everything is reviewed and accepted.

Thanks.

rjd22 commented:
@enov on the PHPUnit problem. Would it be possible to just update the PHPUnit version to 4.3+ in `https://github.com/kohana/unittest/blob/3.3/master/composer.json` ?

rjd22 commented on classes/Kohana/Log.php:
```diff
@@ -107,6 +122,29 @@ public function detach(Log_Writer $writer)
+	 * @param boolean $write_immediately
+	 * @return Log
+	 */
+	public function set_immediate_write($write_immediately = TRUE)
```
Does this need the default `TRUE` in the parameter? I would say that explicitly setting this would make this function more obvious and gives it less redundancy.

enov replied:
I think you're right about explicitly setting this.

enov commented:
Yes, if the tests are finalized the way they are right now, updating the PHPUnit version to 4.3 might do the job.

rjd22 commented on classes/Kohana/Log.php:
```diff
@@ -23,9 +23,24 @@ class Kohana_Log {
 	 * @var  boolean  immediately write when logs are added
 	 */
-	public static $write_on_add = FALSE;
+	protected $write_immediately = FALSE;
```
This can be a private right? Since it can be changed by the `set_immediate_write`

enov replied:
`protected` gives access to classes that extends this class.

rjd22 replied:
@enov the question is why would they need access. Normally I would agree in the sense of CFS but in this case the inner workings are already exposed.

enov replied:
Not sure why would they need, but not sure either why restrict? What do we loose?

rjd22 replied:
Not much ;), If you think extending is important then leave them protected.

acoulton replied:
Every protected instead of private variable is a variable that can only in future be renamed/refactored in a breaking version release. For package code, we should probably lean to private by default unless there's a good reason to make it more visible. After all, if someone has a problem we can easily make it `private` => `protected` in a minor version release. We can't go the other way.

(especially since there's a getter/setter for this, so no real need for anyone extending to have direct access to the variable).

enov replied:
Hmmm, I see. Did not think about refactoring in minor versions. That's a +1 for `private`.

rjd22 commented on classes/Kohana/Log.php:
```diff
@@ -225,4 +240,99 @@ public function write()
+	 *     ));
+	 *
+	 * @uses   Text::interpolate Inserts the replacement values into the message
+	 * @param  mixed  $level    level of message
```
This should only accept `int|string`

enov replied:
https://github.com/php-fig/log/blob/master/Psr/Log/LoggerInterface.php#L116

rjd22 replied:
Wow, ok then leave it there ;)

rjd22 commented on classes/Kohana/Log.php:
```diff
@@ -225,4 +240,99 @@ public function write()
+			throw new Psr\Log\InvalidArgumentException('Undefined level "' . $level . '"');
+		}
+
+		// cast $message to string in compliance with PSR-3
```
No need for this comment. We can see what you're doing ;)

enov replied:
I usually comment excessively, but this is needed to show compliance.

rjd22 replied:
Ok, then leave it there.

enov replied:
:+1:

rjd22 commented on classes/Kohana/Log.php:
```diff
@@ -23,9 +23,24 @@ class Kohana_Log {
+	 * Numeric log level to string lookup table.
+	 * @var array
+	 */
+	protected $_log_levels = array(
```
This can be private? AFAIK

enov replied:
`protected` gives access to classes that extends this class.

rjd22 replied:
Ok leave it protected

rjd22 commented on classes/Kohana/Log.php:
```diff
@@ -225,4 +240,99 @@ public function write()
+	 */
+	public function log($level, $message, array $context = [])
+	{
+		// check if level is available
```
// check if log level exists in the `$this->_log_levels` array.

enov replied:
:+1:

rjd22 commented on classes/Kohana/Log.php:
```diff
@@ -225,4 +240,99 @@ public function write()
+			$context['exception'] = FALSE;
+		}
+
+		// Create a new message
```
Redundant comment.

rjd22 commented on classes/Kohana/Log.php:
```diff
@@ -225,4 +240,99 @@ public function write()
+			'line' => isset($trace[0]['line']) ? $trace[0]['line'] : NULL,
+		);
+
+		// add exception object to message, if available
```
Redundant comment.

rjd22 commented on classes/Kohana/Log/Syslog.php:
```diff
@@ -42,16 +57,38 @@ public function write(array $messages)
+			if (isset($message['exception']))
 			{
-				syslog(Log_Writer::$strace_level, $message['additional']['exception']->getTraceAsString());
+				// convert the level into int level
```
// convert PSR log level into syslog log level

enov replied:
:+1:

rjd22 commented:
@enov a few comments. You are commenting quite excessively. Please check all your comments in the review screen and see if the code is not telling the story itself.

Comments should be used as less as possible since they won't get maintained and will become obsolete when someone starts refactoring.

enov commented:
Will review again my comments, and those that are redundant, I will remove them.

Thanks for taking time reviewing :+1:

Yahasana commented on classes/Kohana/Log.php:
```diff
@@ -225,4 +240,99 @@ public function write()
+		// Grab a copy of the trace and sanitize $context['exception']
+		if (
+		  isset($context['exception']) AND
+		  is_object($context['exception']) AND
```
`is_object($context['exception']) AND` no need

Yahasana commented on classes/Kohana/Response.php:
```diff
@@ -556,10 +556,13 @@ public function send_file($filename, $download = NULL, array $options = NULL)
 
-					// Make sure the logs are written
-					Kohana::$log->write();
+					if (Kohana::$log instanceof Kohana_Logger)
```
why only write it for Kohana_logger?

``` php
Akunamutata_Log extends Psr\Log\LoggerInterface { ... }

Kohana::$log = new Akunamutata_Log;
```

enov replied:
@Yahasana kindly note that this `if` test is to insure that the `$log` can `write()`.

Yahasana replied:
Oops, you are right, there is no 'write' in `Psr\Log\LoggerInterface`. WTF

Yahasana commented on classes/Kohana/Session.php:
```diff
@@ -374,8 +374,14 @@ public function write()
-			Kohana::$log->add(Log::ERROR, Kohana_Exception::text($e))->write();
+			Kohana::$log->error(Kohana_Exception::text($e));
 
+			if (Kohana::$log instanceof Kohana_Logger)
```
same probs as above

enov replied:
same comment as above :stuck_out_tongue_winking_eye: To call `Kohana::$log->write();` we need to be sure that it is a `Kohana_Logger`

enov replied:
You see @rjd22, there is not enough comments to explain the things. I think I need to add some explanation here.

Yahasana commented on classes/Kohana/Log.php:
```diff
@@ -225,4 +240,99 @@ public function write()
+			// remove the call that comes from Psr\Log\AbstractLogger
+			// in order to have consistent file and line elements in log
+			$parent_class_file = (new ReflectionClass(get_parent_class()))->getFileName();
+			if (isset($trace[0]['file']) AND  $parent_class_file === $trace[0]['file']) {
```
ill format `{}`

enov replied:
:+1:

Yahasana commented on classes/Kohana/Log.php:
```diff
@@ -225,4 +240,99 @@ public function write()
+	public function log($level, $message, array $context = [])
+	{
+		// check if level is available
+		if (is_int($level) AND array_key_exists($level, $this->_log_levels))
```
`$this->_log_levels[$level] === NULL` shouldn't happen, replace `array_key_exists($level, $this->_log_levels)` with `isset($this->_log_levels[$level])` for better performance

enov replied:
:+1: yes, `$this->_log_levels[$level] === NULL` shouldn't happen because of `is_int` test, and `isset` will be faster.

Yahasana commented on classes/Kohana/Kohana/Exception.php:
```diff
@@ -145,10 +145,13 @@ public static function log(Exception $e, $level = Log::EMERGENCY)
 
-			// Make sure the logs are written
-			Kohana::$log->write();
+			if (Kohana::$log instanceof Kohana_Logger)
```
```php
if (Kohana::$log instanceof Psr\Log\LoggerInterface)
```

enov committed:
Remove Log::set_immediate_write param default value

Explicitly setting the value would make the function more obvious

Thanks @rjd22

enov committed:
Use isset() instead of array_key_exists() for performance

`$this->_log_levels[$level]` can't be `NULL` because of is_int test

Thanks @Yahasana

enov committed:
Remove redundant `if` test

enov committed:
Fix/remove comments, format code

rjd22 commented on classes/Kohana/Log.php:
```diff
@@ -107,6 +122,29 @@ public function detach(Log_Writer $writer)
+	 * @param boolean $write_immediately
+	 * @return Log
+	 */
+	public function set_immediate_write()
```
@enov Now you removed the variable completely ;). I think it should be 

``` php
public function set_immediate_write($write_immediately)
```

enov replied:
Yeah, my mistake. Working hastily. :neckbeard:

rjd22 commented:
@enov would it be a good idea to update the kohana/unittest to alway use PHPUnit 4.3 at a minimum? That way we can make your tests pass.

enov commented:
@rjd22 @Yahasana I took care of the things you asked. Thanks.

I think I still have a couple of things missing:
- Write test for Text::interpolate
- Support for Object `$level` with `__toString()` that translates to one of the PSR levels - but I am not sure if this is necessary, or if it is an overkill

Meanwhile, @acoulton if you could please review.

@rjd22 

> @enov would it be a good idea to update the kohana/unittest to alway use PHPUnit 4.0 at a minimum? That way we can make your tests pass.

You mean PHPUnit 4.3? Sure, I'll issue a PR, we can merge that if there is no objections to the tests done with `@runInSeparateProcess`.

rjd22 commented:
@enov https://github.com/kohana/unittest/pull/59 This should fix the issue right?

enov commented:
@rjd22 yeah, perfect!

acoulton commented:
@enov I think this is good, nice work! The tests for `Kohana::init` behaviour are a good solution and worth having, I think.

One thing (following to @Yahasana comments) why not have:

``` php
interface WriteableLogger {
  public function write();
}

class Kohana_Log implements Kohana_Logger, WriteableLogger {
//...
}
```

and then everything like this:

``` php
if ($log instanceof Kohana_Logger) {
  $log->write();
}
```

becomes

``` php
if ($log instanceof WriteableLogger) {
  $log->write();
}
```

No need for comments because the reason for the instanceof is very obvious, and much more flexible because someone can define their own PSR logger with a write method without having to also support our attach and detach methods.

enov commented:
@acoulton `WriteableLogger` is great idea! Probably `Kohana_Writeable_Logger` for the name?

acoulton commented:
@enov yeah `Kohana_Writeable_Logger` would work given we're not using namespaces yet.

enov commented:
@acoulton while I have almost implemented `Kohana_Writeable_Logger`, but I am reluctant to push. I see now that `Kohana_Logger` is more of a complete implementation of the observer pattern, and without a `write()`, it would not notify a subscribed writer.

While I see the benefit of having a separate `write()`, `Kohana_Logger` alone or  `Kohana_Log_Writer_Container` alone (I ended up calling it "container", as `attach` and `detach` makes it a mere container) - has little value.

Let me know what you think.

acoulton commented:
@enov well a logger that didn't implement delayed write - eg that wrote to some stream on every log call would only need `attach` and `detach` - but would not need or want to declare an explicit `write` method?

Likewise a logger that only supported deferred output to a single log destination might want `write` but not `attach` and `detach`.

acoulton commented:
@enov thinking about it, maybe the problem is the interface/method name. 

`Kohana_Flushable_Logger` and a `flush` method might make more sense - since probably all loggers can write, it's delayed write/flush that makes the default kohana logger different. But that would be an additional BC break.

If we want to keep the method called `write` then maybe the interface should be more like `Kohana_BufferableWrite_Logger` - I don't think a `flushable` interface makes sense with a `write` method, and the logger may or may actually be buffering depending on the immediate_write option.

enov commented:
Exactly! I was just going to say that... We just need some renaming/refactoring. I was myself thinking also to rename `write` to `flush`. I'll come up with some renamings tomorrow. Thanks!

rjd22 commented:
Ran the tests. It seems that everything is okay now ;)

enov committed:
Test Text::interpolate

The provider and the related test are adapted from
Monolog\Processor\PsrLogMessageProcessorTest

enov committed:
Reset expected log time when testing logging (for HHVM)

HHVM runs Kohana tests slowly, and that's slower with the introduction
of some tests that run on a separate PHPUnit process. We're going to
reset the time within the test.

enov commented:
Added a test for `Text::interpolate`.

@acoulton I am thinking renaming `Kohana_Logger` interface to  `Kohana_Log_Collector` or  `Kohana_Log_Entry_Collector` , and `write()` to `flush()`. That helps emphasizing the role of the interface. It might go hand in hand if we add later a class `Log_Entry` for the log message and an ArrayAccess `Log_Entry_Collection`, similar to the idea you posted here:

https://github.com/kohana/core/pull/563#issuecomment-61961254

> Might it be worth introducing a class (perhaps with ArrayAccess for compatibility) for the log message itself, rather than the simple array that Log::log pushes onto the _messages array? Would allow better typehinting for that property and also for writer methods etc.

enov committed:
Reindex log levels array using explicit integers

Windows has less loglevels which makes the array loose some values. See
https://bugs.php.net/bug.php?id=18090

Also use static to make it class-wide definition, similar to Monolog:
https://github.com/Seldaek/monolog/blob/419ca7ac2562fe43941692ec442ce6d0398dfe10/src/Monolog/Logger.php#L94

enov committed:
Support object log levels

Separate validation and normalization into a function.

enov committed:
Refactor Log_Writer properties

Convert static $timestamp, $timezone, and $strace_level properties to
non-static, add getters/setters. Add default $format property.

enov committed:
Rename `Log::get_level` to `Log::to_psr_level`

`get_level` felt like it's kind of a getter.

enov committed:
Add tests for the new getters setters

Fix trace_level to strace_level

enov committed:
Refactor Kohana_Logger interface to Kohana_Log_Buffer

Kohana_Log_Buffer does not extend Psr\Log\LoggerInterface anymore, it
just provide methods for deffered logging using the observer pattern.

The write() method is now renamed to flush()

enov commented:
Renamed `Kohana_Logger` to `Kohana_Log_Buffer`, without extending `Psr\Log\LoggerInterface`.

@acoulton @rjd22 please review 0825761

rjd22 commented:
Looks good :+1: from me

enov committed:
Simplify interface Kohana_Log_Buffer

Move logic regarding the filtering of specific log levels to the
Log_Writer class.

acoulton commented on tests/kohana/LogTest.php:
```diff
@@ -61,51 +61,360 @@ public function test_attach_attaches_log_writer_and_returns_this()
+				[], // context
+				// expected raw log message array
+				[
+					'time' => time(), // to be reset within the test for HHVM
```
Maybe don't set the time at all in the data provider and just add it in the test message, to make it most obvious?

acoulton commented on classes/Kohana/Log/Writer.php:
```diff
@@ -67,24 +62,202 @@
+	 */
+	public function filter(array $messages)
+	{
+		if (empty($this->levels))
```
This feels a bit backwards - I wouldn't expect $logger->set_levels([]) to result in it logging all messages.

enov replied:
I'll just test for `NULL`. But it's rather `$writer->set_levels()`.

acoulton replied:
I wouldn't expect `NULL` to make it log everything either :) I'd prefer if it was explicit - default $levels to contain all levels.

If it's just that you want to skip filtering for performance if it's not required (?) then if $levels was a map of level => true/false you could do `if ($this->levels === array_filter($this->levels))` to know whether any have been excluded.

acoulton commented on classes/Kohana/Log/Writer.php:
```diff
@@ -67,24 +62,202 @@
+	 * @throws InvalidArgumentException
+	 * @return Log_Writer
+	 */
+	public function set_levels(array $levels)
```
I don't think it's totally clear what type of levels you're accepting here. PSR levels?

enov replied:
Any type, integer levels like `LOG_DEBUG`, `LOG_INFO`, PSR/string levels...

I'll add examples to the docs.

acoulton replied:
Isn't that confusing then if `->get_levels` is always returning an array of PSR levels?

``` php
$levels = ['stuff'];
$writer->set_levels($levels);
if ($writer->get_levels() != $levels) print `eh?`
```

Obviously the logger itself has to accept anything for PSR compliance, but for the writer we should pick a single set of levels and stick to them.

enov replied:
Well, I see that it's a bit confusing. We can add just a note about it instead of sticking with a single set. Otherwise, it may be more confusing that the logger accepts anything, but the writer sticks with a single set?

acoulton replied:
IMO having a getter return different value types to the setter is much more confusing/problematic than this interface (of our own) being stricter than the generic logger interface.

For example if you configured it with `[LOG_DEBUG, LOG_INFO]` etc you might assume elsewhere you could temporarily suppress a level by calling `get_levels`, removing (LOG_DEBUG), and passing back to `->set_levels` but that won't work.

I would make all these methods require PSR LogLevels and throw if they get anything else, so the developer encounters a one-time exception and figures it out rather than getting unexpected behaviour at runtime.

enov replied:
Please note that a bit below, we have `set_levels_range` that needs to accept integer levels for it to convey more its meaning.

acoulton replied:
I don't think it does, I think it can also take PSR levels.

People in the past will /should have been doing `Kohana::$log->attach($writer, Log::CRITICAL, Log::NOTICE)` and in the new implementation they'll do `$writer->set_levels_range(LogLevel::NOTICE, LogLevel::CRITICAL)`.

I don't think the actual value in the constants matters that much. Also forcing them to change their call may help them notice that it's now (correctly) `($min_range, $max_range)` instead of `($max_range, $min_range)`.

acoulton commented on classes/Kohana/Log/Writer.php:
```diff
@@ -67,24 +62,202 @@
+	/**
+	 * Sets the log levels' range that this writer accepts to write
+	 *
+	 * @param array $levels
```
method docs here are wrong.

enov replied:
Yep :+1:

acoulton commented on classes/Kohana/Log/Writer.php:
```diff
@@ -67,24 +62,202 @@
+
+		foreach ($messages as $message)
+		{
+			if (in_array($message['level'], $this->levels))
```
in_array probably isn't the most efficient for multiple messages.

``` php
$include_level = array_fill_keys($this->levels, TRUE);
// Or you could actually store $this->levels as a map of level => include true/false
foreach ($messages as $message) {
  if (isset($include_level[$message['level']])) {

```

rjd22 replied:
@acoulton readability > efficiency. This would make it harder to read imo.

enov replied:
Hmmm... I copied/pasted the stuff from the old `Log::write()` method. I'll look at this again.

acoulton replied:
@rjd22 imo most readable would be `if ($this->write_level[$message['level']])`

acoulton replied:
@enov isn't it annoying when you copy/paste code and people who never cared about it before start picking holes in it :) Sorry about that.

enov replied:
Haha! no problem @acoulton !

acoulton commented on classes/Kohana/Log/Writer.php:
```diff
@@ -11,38 +11,33 @@
+	/**
+	 * @var array log levels that this writer accepts to write
+	 */
+	private $levels;
```
I'm not sure the meaning of `levels` is obvious - same for the various setters/getters. `write_levels`, `include_levels`?

acoulton commented on classes/Kohana/Log/File.php:
```diff
@@ -84,7 +84,9 @@ public function write(array $messages)
-		foreach ($messages as $message)
+		$filtered_messages = $this->filter($messages);
+
+		foreach ($filtered_messages as $message)
```
I might have missed it, but I don't think the filtering within the log writers (any of them) is covered by tests?

enov replied:
I did not have finished yet the testing of the Log_Writer.

acoulton commented:
@enov thanks for all your work on this, I like where it's going. Will you be able to write some documentation on how to upgrade from a 3.3-style logger/writer setup as well?

And sorry if I confused you by adding some comments as commit comments first, I forgot they show up differently / less usefully in the PR conversation so deleted and moved to PR comments.

enov commented:
Thanks @acoulton. Glad to know you like where it is going. I'll add some docs once I finish the tests for the writer. We can then skim through everything, review the docs, and happily see this merged :)

enov committed:
Remove setting the time in provider

enov committed:
Rename the property `levels` to `write_levels`

Same for the various setters/getters.

enov committed:
Fix Log_Writer::set_write_levels_range parameters doc bloc

enov committed:
Make the writable levels explicit

$write_levels is now an array of all levels with the key as the level
and a boolean value TRUE when the level is writable.

enov committed:
Remove ambiguous get_write_levels, add three specific getters

`Log_Writer::get_psr_write_levels_map()`: Gets an array mapping PSR
levels to boolean values with TRUE indicating that the level at the key
is writable

`Log_Writer::get_psr_write_levels`: Gets the PSR log levels that this
writer accepts to write

`Log_Writer::get_int_write_levels`: Gets the integer log levels that
this writer accepts to write

enov committed:
Fix some doc blocks

enov committed:
Add preliminary tests to Log_Writer::filter

enov commented:
I have taken care of all change requests:
- Set the 'time' in test (dataprovider NULL) https://github.com/kohana/core/pull/563#discussion_r43485352 in https://github.com/kohana/core/commit/df8652c64df059f2d4e0a46887629a2945db704d
- $logger->set_write_levels([]) explicit, default to contain all levels https://github.com/kohana/core/pull/563#discussion_r43486184 in https://github.com/kohana/core/commit/dde274d06cfc0226d899921821757104f2775b2b
- **Still accepting different types of levels** but made distinct getters instead of the ambiguous one. https://github.com/kohana/core/pull/563#discussion_r43485486 in https://github.com/kohana/core/commit/14189a683500bd98569e419c36058fc3020d9802
- Fixed some doc blocs https://github.com/kohana/core/pull/563#discussion_r43485502 in https://github.com/kohana/core/commit/e767c0985c4e6e44d2702112be8f81a73a23836e
- Made filtering more efficient, but maybe less readable, relied on map of levels https://github.com/kohana/core/pull/563#discussion_r43485748 again in https://github.com/kohana/core/commit/dde274d06cfc0226d899921821757104f2775b2b
- Renamed `levels` as `write_levels` as well as respective setters/getters, **but still not 100% convinced, maybe `filter_levels`**? https://github.com/kohana/core/pull/563#discussion_r43485846 in https://github.com/kohana/core/commit/c7302401ae134e4b08795e3a303a9272968bb288
- Added some preliminary tests to filter, more to come. https://github.com/kohana/core/pull/563#discussion_r43485989 in https://github.com/kohana/core/commit/38515d0b201bea4899354237ef14297555029cfd

I welcome all comments, my scrutinizers :)

enov commented:
...and maybe we should have a :cake: , been a year :)

enov commented:
It would be nice if you could comment on the last commits, so that I might be able to resume.

@acoulton I am specifically awaiting your comments on https://github.com/kohana/core/commit/14189a683500bd98569e419c36058fc3020d9802 regarding the types of getters/setters of write_level, those are still expecting both integer _and_ psr/string types. Getters and setters are not symmetric anymore, but I left the door open for integer levels for those who prefer to use it, as it seems there is an [open/unresolved issue](https://github.com/php-fig/log/pull/21) on the FIG. Yeah, there might be less than 1% chance for PSR-3 v2, or a follow-up PSR-n, but keeping accepting integer levels in the API might alleviate us from a technical dept in the future. This would also keep consistence with the Log.

Also, cc @rjd22 if you have any reservations.

enov committed:
Simplify filtering API

- Symmetric getters/setters, get what you set
- Simplified variable and methods naming
- Throws InvalidArgumentException when setting duplicate levels
- Remove filtering in range, easily achievable in userland

enov commented:
Please review https://github.com/kohana/core/commit/7dfcc7bb39349c0380537e0316a79b16b8d1d4ef.
Thanks :)

rjd22 commented:
@enov I really don't like the name `filter_map`. It's kind of unclear what its meaning is or what it should do.

enov commented:
Thanks for reviewing @rjd22. What do you suggest?

rjd22 commented:
@enov could you explain what it's task is, exactly? So I can understand it and think of a appropriate name.

enov commented:
`filter_map` is the variable that is actually being used when `filter()`ing. The other `filter` variable only holds the value that is being `set`, so it could `get` back.

`filter_map` could have been computed and used only locally within `filter()`, but that would duplicate some of the code used for validating/normalizing.

rjd22 commented:
@enov what is it filtering? And what is the reason it needs to filter it?

enov commented:
The original `Log_Writer` had the option to limit what it could actually write, by selecting the levels it wants to filter.

enov commented:
I should probably go for `filter_levels` instead of just `filter`?

rjd22 commented:
I would name it something like. `enabled_log_levels`. I would remove the booleans from it and enable them if they are in the list and disable them when they are not.

I would remove the $this->filter parameters
I would remove `get_filter_map` or make it private if needed inside
I would rename `get_filter` to `get_enabled_log_levels`
I would rename `set_filter` to `set_enabled_log_levels`

Also make every function that you don't use/need outside the class private. Making everything public will make future refactoring a problem.

I'm sorry for the changes but please remember that naming is extremely important.

enov commented:
Sorry for the late reply @rjd22. Personally I don't mind, even though I find them a bit verbose :)

I might remove `get_filter_map` altogether :)

acoulton commented:
@enov apologies for the very slow response.

I prefer `writeable_log_levels` to `enabled_log_levels` but either is OK.

I disagree with @rjd22, I think a hash of `level => boolean` is more useful than an array of just the levels that are active. But I would initialise it using the list of defined levels from `Kohana_Log` rather than have a duplicate hardcoded list of them all. 

I don't care too much about this though if you guys disagree, but we should only provide the values in one format - the user can convert from array => map or vice versa if required.

I still increasingly strongly think you should only accept PSR levels when setting the filter (and validate this). All log messages received by the writer already have a PSR log level regardless of input (they're converted in `Kohana_Log::log`). If you force PSR levels for the filter, the developer will almost certainly just reference the defined PSR constant in their code. This has the least possible confusion and is broadly safe whatever operating system they're on and regardless of what actual value PSR assign to the constants in future.

In the worst case if in future PSR switch to integer log levels per RFC 5424 (`LogLevel::EMERGENCY == 0`) but the user is on a system where `LOG_EMERG == 1` you have no idea which kind of integer the user passed in and whether they want to filter emergency or alert messages.

acoulton commented on classes/Kohana/Log.php:
```diff
@@ -10,7 +10,7 @@
+class Kohana_Log extends Psr\Log\AbstractLogger implements Kohana_Log_Buffer {
 
 	// Log message levels - Windows users see PHP Bug #18090
 	const EMERGENCY = LOG_EMERG;    // 0
```
I realise this has always been inconsistent (and always slightly will be for the syslog writer and/or if people are using syslog constants in their own code).

But it makes no sense that for the new logger on some (eg windows) systems `Kohana_Log::EMERGENCY == 1 == \Psr\Log\LogLevel::ALERT`.

In order of preference, we should:
- remove these constants (biggest bc break, best for us)
- make them point directly to the equivalent PSR LogLevel constant
- hardcode the integer values to match the keys of `$log_levels` below (unlikely to break anyone else's code unless they're a windows user relying specifically on `Kohana_Log::EMERGENCY === 1`)

enov replied:
I would prefer going with the last suggestion, to hardcode the integer values. Yes, it removes the inconsistency. And in `Log_Syslog` we're already reverting back to `Log_` values.

acoulton commented:
@enov you should upgrade your :cake: to :birthday: (or possibly even :beers: )

enov commented:
Thanks @acoulton for the review, we should definitely :tada: once this gets merged.

enov committed:
Refactor away log filtering from Log_Writer

In a pattern similar to attaching Log_Writer(s) to the Log,
we can now attach log filters to the writers.

Add Log_Filter_PSRLevel to filter logs by PSR levels.

enov commented:
I have taken away the actual log filtering from the Log_Writer. It can be used maybe elsewhere, not only in the writer. The `Log_Writer` is now `Filter_Aware`, and you can attach as many filters to it as you like, similar to the API you attach writers to the log.

We might only need `Log_Filter_Level` class to be at par, feature-wise, with what already we had. But I have provided some other suggestions, since it now opens new ways of filtering logs. The filters `Log_Filter_Body`, `Log_Filter_File` and `Log_Filter_Exception` could stay if there is consensus about the implementations, or could be removed, and be provided as examples in the docs. `Log_Filter_Union` is there to have an OR union (instead of intersection, AND) when filtering logs with multiple filter requirements.

So much for filtering, I guess. Here's the API:

``` php
$writer->attach_filter(new Log_Filter_Level([LogLevel::EMERGENCY, LogLevel::ALERT]));
```

What we used to do, is to specify the level filtering when attaching the writer to the log:

``` php
$log->attach_writer($writer, Log::EMERGENCY, Log::ALERT);
```

Let me know what you think. Thanks :smile:

rjd22 commented:
@enov I didn't look in depth but as far as I can see it looks good.

enov commented:
Thanks @rjd22.

On 11/13/2015 03:36 PM, Robert-Jan de Dreu wrote:

> @enov https://github.com/enov I didn't look in depth but as far as I 
> can see it looks good.
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/kohana/core/pull/563#issuecomment-156432792.

enov commented:
At this moment, `Kohana_Log_Buffer` interface methods accept a `Log_Writer` which is an abstract class. Being a partial implementation, it is better to be renamed to `Log_Writer_Base`, allowing the introduction of an interface `Log_Writer`.

To remove the confusion, the `Log_File` and the `Log_Syslog` should be renamed to `Log_Writer_File` and `Log_Writer_Syslog` respectively. So basically,

| Current class name | New class name |
| --- | --- |
| Log_Writer | Log_Writer_Base |
| Log_File | Log_Writer_File |
| Log_Syslog | Log_Writer_Syslog |

All those three files would nicely sit in the same directory.

Now, I am not able to proceed with those changes as it breaks the tests, because Koharness gets bootstraped with a `Log_File`. So, I wonder, should I only rename the `Log_Writer` and introduce the interface? Or should I proceed with all the changes and we make Koharness compatible? Or should I just forget about all those things and concentrate on finishing this PR as it is now?

rjd22 commented:
@enov I would start wrapping up the PR. You can always make a new one fixing those names before the release.

I would not make a interface for a abstract class since it already enforces some kind of interface. So I would leave the name `Log_Writer` the same but rename the `Log_File` to `Log_Writer_File` and `Log_Syslog` to `Log_Writer_Syslog`.

acoulton commented:
@enov stop! (in the nicest possible way).

It's getting too hard to follow the changes in this PR already. I like `Log_Filter` but adding the new ones starts to significantly expand the scope of this branch.

I recommend:
- Keep `Log_Filter_Level` - but make it `Log_Filter_PSRLevel` and be strict about the levels you accept (sorry, I'm not going to keep banging on about that - your new design makes it much easier for someone to implement a filter using some other kind of level if they really want to).
- Remove all the other `Log_Filter` implementations - including `Log_Filter_Union` - into a new branch
- Deal with any renaming of `Log_File` etc in a new separate branch
- Change `Log::EMERGENCY` etc to be hardcoded integer values
- Add upgrade notes, documentation etc
- Then we can do a final review of this PR
- And then once this is merged you can make new PRs for your new branches so we can look at them properly.

enov commented:
I guess you're both right. I need wrap up. I guess I got drifted away at some point.

acoulton commented:
it's very easily done :grinning:

enov committed:
Hardcode int log levels as LOG_ consts are inconsistent across OS

Windows users see PHP Bug #18090
Add this note in the Syslog writer as well.

enov committed:
Write better tests for the syslog writing and filtering

enov committed:
Add unit test for Log_File writer

enov committed:
Add logging userguide and upgrading info to the docs

enov commented:
This is ready for final review :)

enov commented:
@acoulton @rjd22 I am hoping to see this reviewed. Is there anything missing? Or should I proceed with squashing commits together?

acoulton commented on tests/kohana/ExceptionTest.php:
```diff
@@ -96,4 +96,108 @@ public function test_text($exception, $expected)
+		$this->assertSame($expected, Kohana::$log->logs[0]['message']);
+
+		// clean up
+		Kohana::$log = $temp;
```
@enov I would move this to the `setUp` and `tearDown` method - as you have it the cleanup won't run if the assertion fails which may cause issues elsewhere.

acoulton commented on tests/kohana/Log/FileTest.php:
```diff
@@ -0,0 +1,216 @@
+	 * @test
+	 * @dataProvider provider_file
+	 */
+	public function test_file(array $log_entries, $expected)
```
Can you add a test that multiple writer instances append properly to the file as expected? (or, that a single instance appends to an existing file with some content which is basically the same).

enov replied:
Tested that a single writer instance appends to an existing file https://github.com/kohana/core/commit/df30f2fb7b95f31b4bddc182cdb6408c56c4eee3

acoulton commented on tests/kohana/Log/FileTest.php:
```diff
@@ -0,0 +1,216 @@
+	{
+		parent::setUp();
+
+		$this->log_dir = sys_get_temp_dir()
```
I think you should remove this directory in tearDown - or better yet use https://github.com/mikey179/vfsStream to test the file writer against a virtual filesystem.

enov replied:
`mikey179/vfsStream` has been used https://github.com/kohana/core/commit/430a1d2e0cdbec8b197a43132b0ec988acababc5

acoulton commented on tests/kohana/Log/WriterTest.php:
```diff
@@ -0,0 +1,196 @@
+	 * @test
+	 * @dataProvider provider_getters_setters
+	 */
+	public function test_getters_setters($property, $value, $exception)
```
I wouldn't usually bother testing setters and getters (apart from testing if they're supposed to throw) - it's enough to test the methods that use the value that's been set IMHO.

enov replied:
Removed `Kohana_Log_WriterTest` https://github.com/kohana/core/commit/304678db4944edfa9811e4f4a37339f8449c697e

acoulton commented on tests/kohana/Log/WriterTest.php:
```diff
@@ -0,0 +1,196 @@
+ * @copyright  (c) 2008-2012 Kohana Team
+ * @license    http://kohanaframework.org/license
+ */
+class Kohana_Log_WriterTest extends Unittest_TestCase
```
It might be better to make this a base class and extend the other writer tests from it? That might reduce some of the duplication of filtering tests, and ensure that they're all doing eg format as expected - I think that's better than explicitly testing the methods of an abstract class.

enov replied:
`Kohana_Log_AbstractWriterTest` is a base class containing filtering tests
https://github.com/kohana/core/commit/304678db4944edfa9811e4f4a37339f8449c697e

acoulton commented on tests/kohana/LogTest.php:
```diff
@@ -61,51 +61,360 @@ public function test_attach_attaches_log_writer_and_returns_this()
+		$logger->flush();
+		$expected_number_of_logs = 1; // expect logs after forced flush
+		$actual_number_of_logs = count($writer->logs); // recalculate actual
+		$this->assertSame($expected_number_of_logs, $actual_number_of_logs);
```
`$this->assertSame(1, count($writer->logs), 'Expect logs written after forced flush');` would be more readable in code and produce a more obvious failure exception message if it fails IMO.

enov replied:
done https://github.com/kohana/core/commit/a279f7e6ea7a7e6aa383a5ed86e1f4b1f4849f55

acoulton commented:
@enov sorry I missed your earlier message. This is looking good I think, there's a few places you could make some small improvements to the tests but it's very close now.

enov commented:
Thanks @acoulton for the review. I will take care of the notes you left.

enov committed:
Run tests that alter the global var Kohana::$log into separate process

enov committed:
Use package mikey179/vfsStream to test Log_File writer

In Log_File, removed the conversion to `realpath` when storing the log
directory, as it is incompatible with streams.

enov committed:
Test writer does not create new log file if exists, only appends

enov committed:
Add AbstractWriterTest to DRY out filter tests

enov committed:
Add `assertSame` failure message and make the code more readable

enov commented:
@acoulton I took care of the notes you left. Some issues though:
- When extending the base for filter tests, PHPUnit's autoload fails. It turned out that the `kohana` folder just under the `tests` folder is all small caps. This needs to be addressed in a future PR so that a PSR-0 autoloader would work. It's working now, because it's named `AbstractWriterTest.php` and that would make it come before the extending test classes, when PHPUnit traverses the filesystem.
- I have completely removed `Kohana_Log_WriterTest`. In case you think any test is needed, I'll just add it back.
- Composer is not downloading from source when Travis is testing on HHVM. I still did not fully investigate why. Is it related to https://github.com/kohana/kohana/pull/92?

Thanks :)

acoulton commented:
@enov this all looks good to me now. I'm not sure about the issue with travis/composer, the build output for 304678d looks normal to me?

enov commented:
Thanks for reviewing again, @acoulton. Yeah, it built normally after I restarted it, most probably Travis/HHVM issue.

acoulton commented:
@enov :beers: :cake: :balloon: :fireworks: and general :tada:

Thanks very much for this!

enov commented:
Wow :flushed: ... merged!

Thank you @acoulton and thank you all who participated: @cyrrill, @arteymix @Yahasana and @rjd22!

rjd22 commented:
@enov Thanks a log for the hard work!!

enov commented:
> @enov Thanks a log for the hard work!!

Haha! Good one, @rjd22, thanks!
